core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala

```diff
@@ -0,0 +1,314 @@
+          // events that aren't in the tag progress table
+          session.writeProgress(id, seqNrTo, tagPidSequenceNr, offset).onComplete {
+            case Success(_) =>
+              log.debug(
+                "Tag progress written. Pid: {} SeqNrTo: {} tagPidSequenceNr: {} offset: {}",
+                id, seqNrTo, tagPidSequenceNr, offset
+              )
+            case Failure(t) =>
+              log.error(t, s"Tag progress write has failed for pid: {} seqNrTo: {} tagPidSequenceNr: {} offset: {}. " +
+                s"If this is the only Cassandra error things will continue to work but if this keeps happening it till " +
+                " mean slower recovery as tag_views will need to be repaired.",
+                id, seqNrTo, tagPidSequenceNr, offset)
+          }
+      }
+      log.debug(s"Tag write complete. ${summary}")
+      flushIfRequired(buffer, tagPidSequenceNrs)
+      pubsub.foreach {
+        log.debug("Publishing tag update for {}", tag)
+        _.mediator ! DistributedPubSubMediator.Publish("akka.persistence.cassandra.journal.tag", tag)
+      }
+
+    case TagWriteFailed(t, events, previousSequenceNrs) =>
+      log.error(t, "Writing tags has failed. This means that any eventsByTag query will be out of date. The write will be retried.")
+      log.debug("Setting tag sequence nrs back to {}", previousSequenceNrs)
+      // wait for the next poll as not to hammer Cassandra
+      context.become(idle(events ++ buffer, previousSequenceNrs))
+  }
+
+  private def flushIfRequired(buffer: Vector[Serialized], tagSequenceNrs: Map[String, Future[Long]]): Unit = {
+    if (buffer.isEmpty) {
+      context.become(idle(buffer, tagSequenceNrs))
+    } else if (buffer.head.timeBucket < buffer.last.timeBucket) {
+      if (log.isDebugEnabled) {
+        log.debug("Switching time buckets: head: {} last: {}", buffer.head.timeBucket, buffer.last.timeBucket)
+      }
+      val (currentBucket, rest) = buffer.span(_.timeBucket == buffer.head.timeBucket)
+      write(currentBucket, rest, tagSequenceNrs)
+    } else if (buffer.size >= settings.maxBatchSize) {
+      log.debug("Batch size reached. Writing to Cassandra.")
+      write(buffer.take(settings.maxBatchSize), buffer.drop(settings.maxBatchSize), tagSequenceNrs)
+    } else if (settings.flushInterval == Duration.Zero) {
+      log.debug("Flushing right away as interval is zero")
+      write(buffer, Vector.empty[Serialized], tagSequenceNrs)
+    } else {
+      timers.startSingleTimer(FlushKey, Flush, settings.flushInterval)
+      log.debug("Batch size not reached, buffering. Current buffer size: {}", buffer.size)
+      context.become(idle(buffer, tagSequenceNrs))
+    }
+  }
+
+  /**
+   * The first time we see a pid we go to Cassandra to see if this if this
+   * tag/pid has been active before.
+   * No other ActorSystem should be reading/updating this row as a pid is owned
+   * by one node.
```

Comment:
The persistent actor recovery reads its part of this table to see if it should replay events to the TagWriter that could be lost.

I don't think I handle the case you're thinking of tho, I think the recovery should send a msg to the TagWriter regardless to say a particular persistenceId has been started on the node in case of the rebalance rebalance.

Reply:
This is now handled and I have a test for it in `EventsByTagRecoverySpec`
It uses two systems and moves it from one to the other and back again, making sure the sequence numbers are continued.

Reply:
Found another race condition here while doing some start/stop testing 

P1 on SystemA has events 1-4 then shuts down.
SystemA P1 tag write progress for 1-2.
SystemA P1 tag write progress for 3-4 is in flight

P1 on System B starts recovery and reads tag progress up to 2 so replays event 3-4 to its TagWriter.
SystemA write progress for 3-4 is complete.
SystemB tag writer persists 3-4 but sees latest tag as sequence nr as 4 so re-writes the same events with sequence numbers 5-6.

So rather than clearing it I think recovery should first send the progress state to the TagWriter that its basing its recovery from.